posts/2022-10-06-experimenting-with-voting-schemes.adoc

```diff
@@ -0,0 +1,400 @@
+
+== Implementation
+
+https://github.com/agoetschm/voting-helios-scala[This repo] contains a simplified implementation of Helios without any dependencies: the necessary discrete algebra objects and cryptographic primitives are implemented from scratch. Additionally, the code aims to be as clear and simple as possible in order to show how the different parts of the scheme are put together. The strong typing of Scala helps to give expressive signatures to each function and make their role explicit.
+
+=== Structure
+We will go over the different layers composing the code in a bottom-up approach, going from the basic building blocks up to the final voting scheme.
+
+The lowest layer of this implementation is the https://github.com/agoetschm/voting-helios-scala/tree/develop/src/main/scala/algebra/[algebra] package. It contains the logic about https://mathworld.wolfram.com/CyclicGroup.html[cyclic groups] and https://mathworld.wolfram.com/FiniteField.html[finite fields] on which the encryption is based. 
+[source,scala]
+----
+trait Domain[Z: Integral, F <: Field[Z], G <: Group[Z, F], Gen <: Generator[Z, F, G]]:
+  val base: G
+  val exponent: F
+  val generator: Gen
+----
+The https://github.com/agoetschm/voting-helios-scala/tree/develop/src/main/scala/algebra/Domain.scala[Domain] trait englobes a discrete group `base`, which is used later as the ElGamal public key space, with a finite field `exponent`, corresponding to the private key space. The `generator` projects elements of the `exponent` field on a subgroup of the `base` group. Note that `Domain` is "data type agnostic": the type parameter `Z` can be any type representing an integer - such as `Int`, `Long` or `BigInt` - in order to handle different sizes of numbers.
+
+On top of this is built the https://github.com/agoetschm/voting-helios-scala/tree/develop/src/main/scala/crypto/[crypto] package. It contains an implementation of the https://github.com/agoetschm/voting-helios-scala/tree/develop/src/main/scala/crypto/ElGamal.scala[ElGamal encryption scheme] and the https://github.com/agoetschm/voting-helios-scala/tree/develop/src/main/scala/crypto/ProofOfKnowledge.scala[proofs of knowledge] needed for Helios.
+[source,scala]
+----
+trait EncryptionScheme[PT, CT, PK, SK, D, R]:
+  val gen: (D, () => R) => (PK, SK)
+  val enc: (PT, PK, () => R) => CT
+  val dec: (CT, SK) => PT
```

Comment:
Suggestion; I would either add a legend describing the different type parameters or, alternatively swap the abbreviations for descriptive names. (`PT` -> `Plaintext`, ... `D` -> ?, `R` -> ?)